core/src/main/java/com/oath/oak/OakDetachedReadKeyBufferSynced.java

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright 2018 Oath Inc.
+ * Licensed under the terms of the Apache 2.0 license.
+ * Please see LICENSE file in the project root for terms.
+ */
+
+package com.oath.oak;
+
+import java.util.function.Function;
+
+/**
+ * This class is used for when a detached access to the key is needed with synchronization.
+ * It extends the non-synchronized version, and changes only the transformBuffer() method to perform synchronization
+ * before any access to the data.
+ * It is used by non-stream iterators that iterate over the keys (KeyIterator and EntryIterator).
+ */
+class OakDetachedReadKeyBufferSynced extends OakDetachedReadKeyBuffer {
+
+    private final MemoryManager memoryManager;
+
+    OakDetachedReadKeyBufferSynced(MemoryManager memoryManager) {
+        super();
+        this.memoryManager = memoryManager;
+    }
+
+    @Override
+    protected <T> T transformBuffer(Function<OakAttachedReadBuffer, T> transformer) {
+        // Internal call. No input validation.
+        // No access is allowed once the memory manager is closed.
+        // We avoid validating this here due to performance concerns.
+        // The correctness is persevered because when the memory manager is closed,
+        // its block array is no longer reachable.
+        // Thus, a null pointer exception will be raised once we try to get the byte buffer.
+        memoryManager.readByteBuffer(key);
```

Comment:
OK, I get deeper into it now and I understand. You use this memoryManager.readByteBuffer(key) in order to see that memoryManager is not released and this access is outdated. This is legitimate. But this check is necessary also in "unsynced" version of the class, that access can be long time outdated and lead to not correct answer. This is not a synch this is validity of access.

memoryManager.readByteBuffer(key) is invoked on every next strep, but as you did correctly needs to be invoked also on access to see that the access is not outdated. BUT memoryManager.readByteBuffer(key) makes much more then checking validity we may have a slimmer check.

Finally, I currently see no reason in detached-key-sync/unsync. As I see it they are absolutely the same. When we will add keys deletion and reusage we will think what to do.